### ISSUE TITLE
Incorrect interface used for IP route despite setting IP_AUTODETECTION_METHOD

### ISSUE THREAD

**Issue by 2rs2ts** [projectcalico/calico#1480]
## Expected Behavior

When running calico-node on an EC2 instance with two ENIs attached to it, and configuring calico to use the second interface (the one which is not the default interface given by EC2 automatically), the routes created by calico should use that second interface instead of the default one.

## Current Behavior

Although calico does configure the node in etcd to have the right IP address, the IP route shows that the wrong interface is being used.

## Possible Solution

Our theory is that, because the two ENIs are necessarily in the same subnet, bird is getting confused by the fact that there are multiple default routes for the same subnet (one for each interface.) Linux networking is not my domain of expertise, but given the routes that we see via `ip ro sh` I can't think of any other source for the bug.

## Steps to Reproduce (for bugs)

I will try to be thorough in replicating our setup while still trying to keep things minimal, so bear with me. It may be possible to skip a lot of the EC2 boilerplate here, but I don't have the time to try to whittle this down any further than I already have.

1. Set up the EC2 VPC.
    1. Create a VPC in any region (we used `us-west-2`.) Enable DNS hostnames and DNS support. For the subnet, we used `10.42.0.0/19`.
    2. Create a Route53 private hosted zone in that VPC, and an SOA record for that hosted zone. We used `.bt.local` for our hosted zone's name.
    3. Create a public subnet and a private subnet, both in the same AZ/ We used `10.42.0.0/22` for the public subnet and `10.42.16.0/22` for the private subnet, and put both in `us-west-2a`.
    4. Create an IGW and a route table that routes all traffic on the public subnet through it.
    5. Create a NAT Gateway and a route table that routes all traffic on the private subnet through it. (We actually use a NAT box we control, but the effect is the same.)
    6. Create the following security groups:
        1. A security group that allows all traffic within the VPC CIDR block.
        2. A security group that allows ingress ssh traffic from the internet, and all egress traffic to the internet.
        3. A security group that allows all traffic within what will become the calico IP pool. (We used `192.168.64.0/18`.)
2. Spin up some instances. Unless otherwise stated these should be all based on the publicly available Debian Jessie AMI.
    1. A bastion host, on the public subnet and with a public IP, so you can get into the other instances. Attach the first two security groups from before to this instance.
    2. An etcd server for calico to use, on the private subnet. Install and run etcd 3.1.0 on it. Attach the second security group from before to this instance.
    3. Two servers to act as the calico nodes, again on the private subnet. These should be based on the publicly available Debian Stretch AMI instead (in case it matters, I don't want to omit the difference in OS release.) Attach the second and third security groups from before to these instances.

        In our case we are running calico in docker via kubernetes, so if you want to be thorough, follow these substeps:

        1. Install docker 1.13.1, kubelet 1.7.0, and kube-proxy 1.7.0 on these nodes.
        2. Spin up another server (back to Debian Jessie) for the kubernetes control plane and etcd cluster. Attach the second security group from before to this server. Install kube-apiserver 1.7.0, kube-controller-manager 1.7.0, kube-scheduler 1.7.0, and etcd 3.1.0. Hook all that up and create an API token for kubelet to use as well as an SSL cert for the apiserver to use.

        However I don't think this is strictly necessary to do in order to reproduce what we are seeing. I haven't tried doing this with calico-node running directly on the host or at least in docker without kubernetes, but I believe it should behave the same way.
3. Get the ENIs set up.

    Create two ENIs, one for each server running calico-node, and put them in the same subnet as the servers themselves. Make sure source/dest checking is off, and attach the second and third security groups from before to each ENI.

    In our case we have some udev rules set up on the servers. We've vendored the scripts from bootstrap-vz, which you can look at [here](https://github.com/andsens/bootstrap-vz/tree/97378befced235b9c212a3bfb362e929cf1d36ca/bootstrapvz/providers/ec2/assets/ec2). Because these scripts rely on `/etc/sysconfig/network-scripts` existing, create that directory on the calico-node servers before installing these scripts.

    We also have a script in `/etc/network/if-up.d/` you should add as well:

    ```bash
    #! /bin/sh -e

    # only configure additional attached interfaces
    if [ "$IFACE" = "lo" ] || [ "$IFACE" = "eth0" ] ; then
      exit 0
    fi

    # ignore meta entries (ifup -a)
    if [ "$ADDRFAM" = "meta" ]; then
      exit 0
    fi

    if [ ! -f "/etc/sysconfig/network-scripts/route-${IFACE}" ] ; then
      exit 0
    fi

    while read line ; do
      /sbin/ip -4 route add $line 2>&1
    done < "/etc/sysconfig/network-scripts/route-${IFACE}"
    ```

    >If you are following along with the kubernetes steps we took, add a route53 record for each of these ENIs. These records will be the node names given to kubelet and to calico-node. In our case we did `kube-worker01.bt.local` and `kube-worker02.bt.local`.

    Once you have that all set, attach the ENIs respectively to each calico-node host.
4. Start calico-node, making sure to add the following settings:
    * `IP_AUTODETECTION_METHOD`: `interface=eth1`
    * `IP`: the IP address of eth1
    * `NODENAME`: the IP address of eth1 (or if you were doing the kubernetes steps above, the DNS names for the ENIs.)

    >If you're following along with the kubernetes steps then do this by starting the kubelet and kube-proxy (make sure they're configured to talk to the apiserver and they trust the cert you used for that server) and then launch calico as a kubernetes daemonset. I'll share the config details in the Conext section of this bug report.

    Verify with calicoctl that the IP addresses of the extra ENIs are the ones registered for the calico nodes. If that's not the case, then my instructions were not good enough, and you should let me know.
5. Check the IP routes on the calico-node servers (`ip r sh table all`)

    If you see something like `192.168.69.192/26 via 10.42.18.190 dev eth0 proto bird` in there, congratulations! You reproduced the bug. It _should_ have said `eth1` instead.

## Context

I included the bird config, container logs, daemonset configuration, and IP configuration in this gist: https://gist.github.com/2rs2ts/eba5900a81edd575a22a20ca0c25b697 The evidence of the issue can be seen on the routing information, it should say `eth1` for the bird route but it says `eth0`.

Sorry, no `calicoctl node status`, I'm running calico as a daemonset in k8s, so I don't think it's possible.

## Your Environment

* Calico version: 2.6.3
* Orchestrator version (e.g. kubernetes, mesos, rkt): kubernetes 1.7.0, docker 1.13.1
* Operating System and version: Debian Stretch for the calico-node servers, Debian Jessie for everything else

**Comment by 2rs2ts** [comment]
P.S. I have the environment in which this issue is manifesting still running but in a day or two I will have to get rid of it to trim our AWS bill down, so if you want any more debug output please let me know ASAP before I tear this thing down.

**Comment by tmjd** [comment]
Thanks for the report and all the information. I'm sure it took a while to type this up and collect the info, it is appreciated.

You should be able to still run `calicoctl node status` with calico as a daemonset (though you probably need to use `sudo`).

I was looking at the eth0 & eth1 interfaces and they are in the same subnet, right? Is that what you expected?  I'm wondering if you put them in different subnets if the routes would happen as you expect. This is kind of along my line of thinking when we were chatting in slack that bird is picking the route/interface based on the linux standard routing.

**Comment by bkochendorfer** [comment]
It is expected for our deployment that the two interfaces are in the same subnet. This is obviously somewhat unusual and we acknowledge that. However, I think we expected the calico service to write through to the bird config to specify a particular interface.

**Comment by 2rs2ts** [comment]
>You should be able to still run calicoctl node status with calico as a daemonset (though you probably need to use sudo).

From where? 😅 I tried using the calico/ctl pod, and I also tried running it from the etcd host. Both say that the calico process is not running, which makes sense, because calico runs on neither of those.

**Comment by tmjd** [comment]
From one of the hosts running calico-node, if you can, the easiest way is to download the binary and make it executable. It can be done as a pod but requires mounting in /proc and /var/run/calico (I believe).

**Comment by 2rs2ts** [comment]
Oh, didn't even think of that. Yeah here is the output for both servers.

```
Calico process is running.

IPv4 BGP status
+--------------+-------------------+-------+------------+-------------+
| PEER ADDRESS |     PEER TYPE     | STATE |   SINCE    |    INFO     |
+--------------+-------------------+-------+------------+-------------+
| 10.42.18.190 | node-to-node mesh | up    | 2017-12-05 | Established |
+--------------+-------------------+-------+------------+-------------+

IPv6 BGP status
No IPv6 peers found.
```

```
Calico process is running.

IPv4 BGP status
+--------------+-------------------+-------+------------+-------------+
| PEER ADDRESS |     PEER TYPE     | STATE |   SINCE    |    INFO     |
+--------------+-------------------+-------+------------+-------------+
| 10.42.16.78  | node-to-node mesh | up    | 2017-12-05 | Established |
+--------------+-------------------+-------+------------+-------------+

IPv6 BGP status
No IPv6 peers found.
```

**Comment by tungdam** [comment]
I have the same problem today. Calico-node uses docker bridge interface instead of eth0 that i specified by using IP_AUTODETECTION_METHOD=interface=eth0. 

**Environtment:** 
 quay.io/calico/node:v2.6.3.  
 Debian Jessie.

**Comment by tmjd** [comment]
@2rs2ts I chatted with @caseydavenport about this and he confirmed that the decision on how the traffic will be routed is determined by the linux routing table.  The BGP traffic will be over the specified ones but the routes are not forced for the dataplane traffic.

@tungdam I've never seen the docker bridge interface get selected. But one thing to make sure you are aware of when using IP_AUTODETECTION_METHOD that if calico has previously been deployed then that IP_AUTODETECTION_METHOD is not honored as the node resources have already been created.  To have it do update the previously created resource you need to also set `IP=autodetect`.

**Comment by tungdam** [comment]
Thanks. It looks like using IP=autodetect helps:

`Dec 13 14:28:23 cd1v docker[9576]: Using autodetected IPv4 address 10.101.0.51/22 on matching interface eth0
Dec 13 14:28:23 cd1v docker[9576]: WARNING: Calico node 'cd1v' IPv4 address has changed:
Dec 13 14:28:23 cd1v docker[9576]: -  This could happen if multiple nodes are configured with the same name
Dec 13 14:28:23 cd1v docker[9576]: -  Original IP: 172.18.0.1
Dec 13 14:28:23 cd1v docker[9576]: -  Updated IP: 10.101.0.51`

It was like this when we didn't set the IP env:
`Dec 13 14:28:03 cd1v docker[6273]: WARNING: Unable to confirm IPv4 address 172.18.0.1 is assigned to this host
Dec 13 14:28:03 cd1v docker[6273]: No AS number configured on node resource, using global value`

**Comment by 2rs2ts** [comment]
Thanks @tmjd. Naive question: does that make a resolution fundamentally impossible?

**Comment by tmjd** [comment]
TBH I'm not sure what all options you have.
A couple options that may work:
- You could do like I suggested and use a different subnet for those interfaces
- Making sure that the default route on your system for that network is over the interface you want traffic to go over would probably do what you want (but the the traffic you don't want going through that interface may go it then).
- (Total speculation) you may be able to do some custom bird configuration but Calico does not expose options to do that

**Comment by tmjd** [comment]
I'm going to close this issue as I don't think there is anything left to do here as Calico is working as expected.  If there is a new feature that you would like to request please open a new issue for that.

**Comment by mlushpenko** [comment]
@tmjd we have a similar situation to this issue - nodes in public and private zones have two interfaces (management and production) + k8s masters in separate zone where only management interface is available. 

Hostnames that we are currently using for kubernetes/calico setup are resolved to management interface and IP of that interface we can see in the routes that calico distributes over BGP like:

`pod_network` via `management_interface` dev tunl0

But communication between public and private is allowed only via production interface, so masters can talk to nodes in public or private zone via management interface, but nodes between themselves are allowed to talk only via production interface. Is it too confusing? 😃 I added a [picture](https://www.draw.io/?lightbox=1&highlight=0000ff&edit=_blank&layers=1&nav=1&title=calico.xml#R7VtLc6M4EP41Pm4KEOJxnHgmu4edKldlq3bnKINsa4ORC%2BTEya9fCQRGD69JDE6ljH0waj0s6fta3WqJGZhvD78XaLf5SVOczTwnPczA95nnuW7s8B8hea0lUezWgnVBUlnoKHgkb1gKZb31nqS4VAoySjNGdqowoXmOE6bIUFHQF7XYimbqv%2B7QGhuCxwRlpvRvkrKNHAV0jvI%2FMFlvmn92HZmzRMnTuqD7XP7fzAOr6lNnb1HTlixfblBKXzoi8GMG5gWlrH7aHuY4E3PbTFtd7%2BFEbtvvAuesTwUQ1jWeUbbHTZerjrHXZjKeccEIn5s%2F0RJnC1oSRmjOs5aUMbqdgfumwLeMrEUGozsu3bBtxhMuf%2BSj3InGtoe14MvdEpUkuTuIHFbQJzynGS2qPwPLCPp8msH9imRZR76KEpwkXM5bSAkfXScPo2AZBKIOzZkkk5jKKt0p51QfLkeyowlvB%2FPMezkJfBz4cHIm3RYfzntMt5gVr7yIrBCGsK4iKQ8asrwcCdSINh3uNDIkKbtuWz7Cxh8kcnYUfccADaec0DKZ05z%2F3Fe0xKKKo8KDD4T9I8R3AZTJX1WySS1wQXinxFRVdfmUFK%2FdGiLdqSKSep2SoYJ9E7opJj5DJadAI34gWdOXfzFjrxJDtGeUi2jBNnRNc85AKphVU0onznwexw8PbU6jtuAdtKgnUczc%2F3OATzTdFwlWVYgPZI1ZV2ZypcAZYuRZbf8S5D3%2F8%2FSX13myIRF5S1Cro6rCKcRR6ttVOA5TJww%2FWYU9qKqwF1lUGJoqHAygwi44D%2BRRgQUoLxvC8OMOVTx84bhomOmwxNXHsrLW9smERS4bn4kIaBwCiUgADUDcyLKoes4QkHgTJCYk0AeqnQtMJRkRE3fCxMTE9ZxPxKSHmvAK3HXH5%2BFA5a7251fkICA08QmSCC9XFuuCcLRKLgHCYpZC7lpiZyDFCVSQwtgECVgwAgNA1MO4DAjRV1QhHZ3Ivx46pg%2F3E%2BV8i7kVfefFxChXAg4dMz5cpoJT4JK8oWVVQEzVjvLaVdfg%2FQx%2BF5PHneqynmS3M5cZXrGTrmDJyUDy9V%2BV9%2F2b%2F26X%2BiJkAic6qzejQQMNaBYFTfdJ5UJP0LgqNDalaWM6Q0MTTGanl28QqQub71xvYesRXJogMtXomhC5vgnJB0JHzfOv5rmNECnhIZ5IUblpnfNPCgq5oweFJKDdmNC1QkLRpHR9lA5o66JrideOpXTxVSH6iu64YbUsobjRlkRngue9Fuua8ASDW6zOsYVps6q827BasWm1XMdOhN5mS1ZdiN3OkUB%2Bg2Kz%2BkaB2kTdLVnrSA6jIQh9lYmNQ9M0VA%2FGaKhiWTuefsTr4c42JyUZyZ9mWqBxXNIE1eci0ugLEBRfuyVeVt93EM1kUM9z0QuJBlU7AvSAZ1%2BeRe6Zhobjmdcjxj2dyffy8bSgHrCEjsY6k7c5DZOZGsRMBaaZunRzZVd6TztXgQ78qJU609CAq0ePWP90I6DvAgLDO%2FVOgG85gh7rTkCfEOcEZd8DHle73GGzBWNd7oCD2ALnVJCtTdx8kC240A70Vs1pL%2FD19gK%2BZoUDPRzR15zrDY25GTDDuYv9MuN88Zw3sW7orLuJg0f9wguMPWMxd8cKQFnCt4uCPCOGbxoT%2FZw%2BcCyY%2BCNhAsyYrXKH4pZhOX%2FO0V4DGxwWM5TxFJUz8SJEKTy7m0QEaNbD4oi269vggJi7wxqQnKY3qiAwUi%2B0Qs%2BCx2gKYt79unU8XC1od108ht%2BpTVG72andWmwnwoVuPojVI6EPny152iFV37Mljh567RSTSnmyw%2FpuIlTe2eMPdYMf3UKA0Q9M3YnTYx%2BYtkcXDanDD4aiXa2hvqHo95I6jMdlte8ZltOg%2BXRK1sfg6i%2BPjPfiKk8eX22uMT%2B%2BPw5%2B%2FAc%3D)

Any idea how to make calico work with such setup? Routes for the lines with green checkmark are present. I guess the only way to make it work is to add manual routes on nodes like:

`managemnt_interface_in_public_zone` via `prod_interface` on the node in private zone
`managemnt_interface_in_private_zone` via `prod_interface` on the node in public zone

Or there is much better way to do it? If we set up calico for prod interfaces, then communication with masters will be broken. If we set up calico for managemnt interface, then communication between nodes in public/private is broken.
